core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -561,7 +561,7 @@ private[deploy] class Master(
     apps.filter(_.state == ApplicationState.UNKNOWN).foreach(finishApplication)
 
     // Update the state of recovered apps to RUNNING
-    apps.filter(_.state == ApplicationState.WAITING).foreach(_.state = ApplicationState.RUNNING)
+    apps.filter(_.coresGranted > 0).foreach(_.state = ApplicationState.RUNNING)
```

Comment:
Shall we also check the `state` to guarantee only the `WAITING` state and `coresGranted > 0` can be shifted to `RUNNING` ?

Reply:
Correct me if I'm wrong, but isn't the app state should be RUNNING as long as it gained an executor? not matter whatever state before.

Reply:
My only concern is that there might be some corner cases to transfer wrong state to `RUNNING` (though I cannot imagine).

Reply:
> Shall we also check the state to guarantee only the WAITING state and coresGranted > 0 can be shifted to RUNNING ?

Not necessary. When we run into `completeRecovery()`, an app can only be either `UNKNOWN` or `WAITING`. And, all `UNKNOWN` apps have been filtered out when we reach here. So, they are definitely in `WAITING` state at this time. See: 

https://github.com/apache/spark/blob/d9e4cf67c06b2d6daa4cd24b056e33dfb5eb35f5/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L560-L563

Reply:
I'm thinking that there may be a corner case where `_.coresGranted > 0` can not work. For example, in dynamic allocation mode, app A is running on HA Standalone. And master just happens to change during the time app A has released all of its executors, which result in  `coresGranted=0`. Then, app A can't recover to RUNNING but remain in WAITING at this time. @zuotingbing  @jerryshao

Reply:
So this brings out another question, can we say that this application is `WAITING` or `RUNNING` if by judging the executors? Obviously when dynamic allocation is enabled, no executor doesn't mean the application is waiting.

Reply:
So this brings out another question, can we say that this application is `WAITING` or `RUNNING` if by judging the executors? Obviously when dynamic allocation is enabled, no executor doesn't mean the application is waiting.

Reply:
Maybe, the better way to mark an app as `WAITING` is the time the bind driver for that app waiting for an executor to launch on. Once the driver set up, the app is `RUNNING`.

Reply:
> app A has released all of its executors, which result in `coresGranted=0`. Then, app A can't recover to RUNNING but remain in WAITING at this time. 
Why app A release all of its executors? Does app A finished? Or shall we change the state to Running when app A grant executors again?

Reply:
In a Dynamic Executor Allocation enabled scenario, executor number can be 0 in idle state (when min executor number is set to 0). So in this scenario, we can hardly say that application is waiting (and it is not finished).

Reply:
I search for "ApplicationState.RUNNING" in the whole project, only be set in function `org.apache.spark.deploy.master.Master#allocateWorkerResourceToExecutors`, which means if app gained 0 executors it should be the init state(ApplicationState.WAITING)

Reply:
>For example, in dynamic allocation mode, app A is running on HA Standalone. And master just happens to change during the time app A has released all of its executors, which result in `coresGranted=0`. Then, app A can't recover to RUNNING but remain in WAITING at this time. 

After debug, i found when the app gain executors again, the app can recover to RUNNING from WAITING. So it may be OK in master changed scenario that if app gained 0 executors we set it to WAITING state as usual as an app init do.

Reply:
Hmmm...with the current code logic, I'd say yes that an app could always update to RUNNING from WAITING whenever it gains an executor. Then, the WAITING status, here, exactly means that **the app is waiting for at least one executor**. As a result, an app in Standalone with dynamic allocation could change its status from RUNNING/WAITING to WAITING/RUNNING back and forth. But, usually, it doesn't make sense. 

The argument here, I think, is that how do you define the WAITING status for an app in Standalone.  If you say, an app should be in WAITING if it does not own  any executors at the moment. Then, current code logic and PR change would be fine. But, if you say, an app should be in WAITING if it hasn't launched its driver. Then, current code logic may be wrong.

Reply:
> But, if you say, an app should be in WAITING if it hasn't launched its driver

Hmmm... maybe you make it extremely complicated, there is no synchronous causal link between the Driver state and the App state, when driver is active, app can be WAITING or RUNNING.

Reply:
> As a result, an app in Standalone with dynamic allocation could change its status from RUNNING/WAITING to WAITING/RUNNING back and forth

RUNNING to WAITING only exists when Master has changed. Let us go to the beginning. Why we set app to WAITING, then change WAITING to RUNNING during the recovery time? **Shall we keep the state before and just filter out the UNKNOWN app that didn't respond during the recovery time?**

Reply:
From the code, it looks like only when executor is registered, then the state will change to `RUNNING` in master UI (the code there is quite old). But I think it is not valid to use registered executor numbers to judge the state of application:

1. When dynamic allocation is enabled, executors can be 0, we cannot say this application is waiting because it will ramp up soon.
2. Some driver only tasks can be run without executors, we cannot say this application is waiting.

So as a thorough fix, I would suggest to revisit all the changes related to application state, to see whether it is meaningful as for now, rather than just a point fix here.

Reply:
Yes the point is when and why app is set to WAITING state(besides the app init).  
`case MasterChangeAcknowledged(appId) =>
      idToApp.get(appId) match {
        case Some(app) =>
          logInfo("Application has been re-registered: " + appId)
          **app.state = ApplicationState.WAITING**
        case None =>
          logWarning("Master change ack from unknown app: " + appId)
      }

      if (canCompleteRecovery) { completeRecovery() }`

planB: Maybe when master changed ,we just keep the state before(do not need to set app to WAITING) and filter out the UNKNOWN app that didn't respond . I will update code for planB

Reply:
Further learning the logic of recover app when master has changed, i found that in function RegisterApplication , an app is created and then stored in persistenceEngine:  `persistenceEngine.addApplication(app)` as WAITING state. When master has changed, recover app from the stored data and register application again with WAITING state. The current fix maybe right(after recover,the app is on WAITING with init state, then change to RUNNING as long as it gained an executor), otherwise the stored app data in persistenceEngine is wrong.